src/vos/vos_io.c

```diff
@@ -2458,6 +2489,10 @@ vos_update_end(daos_handle_t ioh, uint32_t pm_ver, daos_key_t *dkey, int err,
 		goto abort;
 	}
 
+	if (dtx_is_valid_handle(dth) && dth->dth_local) {
+		err = vos_insert_oid(dth, ioc->ic_cont, &ioc->ic_oid);
```

Comment:
Will local TX touch punch logic? if not, please add assert() at the beginning of vos_obj_punch(), then all subsequent punch related logic will not handle local TX cases.

Reply:
I do not see the relationship between the no need for "touch" (I assume you mean any changes to the punch logic needed specifically for local TX) and the notion that punches are out of the scope for local TX. There is also a simple test for local TX making use `vos_obj_punch()` that works flawlessly so far.

Local TX has to support the punch operation. As far as I am concerned a punch is just a "special value" no need for any special logic to handle it for local TX.

Am I missing something?

Reply:
If punch is in the scope of local TX, why only record the cont/obj that are touched by the local TX for update? Why not for punch?

Reply:
Good catch. Fixed.